server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,580 @@
+// such as "member1=http://127.0.0.1:2380".
+//
+// The final returned string has the same format as "--initial-cluster", such as
+// "infra1=http://127.0.0.1:12380,infra2=http://127.0.0.1:22380,infra3=http://127.0.0.1:32380".
+func JoinCluster(lg *zap.Logger, durl string, cfg *DiscoveryConfig, id types.ID, config string) (cs string, rerr error) {
+	d, err := newDiscovery(lg, durl, cfg, id)
+	if err != nil {
+		return "", err
+	}
+
+	defer d.close()
+	defer func() {
+		if rerr != nil {
+			d.lg.Error(
+				"discovery failed to join cluster",
+				zap.String("cluster", cs),
+				zap.Error(rerr),
+			)
+		} else {
+			d.lg.Info(
+				"discovery joined cluster successfully",
+				zap.String("cluster", cs),
+			)
+		}
+	}()
+
+	return d.joinCluster(config)
+}
+
+type discovery struct {
+	lg           *zap.Logger
+	clusterToken string
+	memberId     types.ID
+	c            *clientv3.Client
+	retries      uint
+	durl         string
+
+	cfg *DiscoveryConfig
+
+	clock clockwork.Clock
+}
+
+func newDiscovery(lg *zap.Logger, durl string, dcfg *DiscoveryConfig, id types.ID) (*discovery, error) {
+	if lg == nil {
+		lg = zap.NewNop()
+	}
+	u, err := url.Parse(durl)
+	if err != nil {
+		return nil, err
+	}
+	token := u.Path
+	u.Path = ""
+
+	lg = lg.With(zap.String("discovery-url", durl))
+	cfg, err := newClientCfg(dcfg, u.String(), lg)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := clientv3.New(*cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &discovery{
+		lg:           lg,
+		clusterToken: token,
+		memberId:     id,
+		c:            c,
+		durl:         u.String(),
+		cfg:          dcfg,
+		clock:        clockwork.NewRealClock(),
+	}, nil
+}
+
+// The following function follows the same logic as etcdctl, refer to
+// https://github.com/etcd-io/etcd/blob/f9a8c49c695b098d66a07948666664ea10d01a82/etcdctl/ctlv3/command/global.go#L191-L250
+func newClientCfg(dcfg *DiscoveryConfig, dUrl string, lg *zap.Logger) (*clientv3.Config, error) {
+	var cfgtls *transport.TLSInfo
+
+	if dcfg.CertFile != "" || dcfg.KeyFile != "" || dcfg.TrustedCAFile != "" {
+		cfgtls = &transport.TLSInfo{
+			CertFile:      dcfg.CertFile,
+			KeyFile:       dcfg.KeyFile,
+			TrustedCAFile: dcfg.TrustedCAFile,
+			Logger:        lg,
+		}
+	}
+
+	cfg := &clientv3.Config{
+		Endpoints:            []string{dUrl},
+		DialTimeout:          dcfg.DialTimeout,
+		DialKeepAliveTime:    dcfg.KeepAliveTime,
+		DialKeepAliveTimeout: dcfg.KeepAliveTimeout,
+		Username:             dcfg.User,
+		Password:             dcfg.Password,
+	}
+
+	if cfgtls != nil {
+		if clientTLS, err := cfgtls.ClientConfig(); err == nil {
+			cfg.TLS = clientTLS
+		} else {
+			return nil, err
+		}
+	}
+
+	// if key/cert is not given but user wants secure connection, we
+	// should still setup an empty tls configuration for gRPC to setup
+	// secure connection.
+	if cfg.TLS == nil && !dcfg.InsecureTransport {
+		cfg.TLS = &tls.Config{}
+	}
+
+	// If the user wants to skip TLS verification then we should set
```

Comment:
It's a good point although it's just a nit. Thanks @spzala 

Actually there are lots of such inconsistencies in the repo. I'd like propose the following rule:
If the comment is a complete sentence, then the first letter should be in upper case, otherwise lower case. We can do some cleanup in separate PR. What do you think? cc  @serathius @ptabor

Reply:
I just updated the comments only in this file per this rule.

Reply:
@ahrtr sounds good, appreciate this discussion. We follow this convention https://go.dev/doc/effective_go#commentary so I think we should clean up accordingly. Ideally, we should always have a complete sentence. Let's have this discussion separately and clean up. Thanks!